[dtensor] refactor redistribute and fix uneven sharding redistribution

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115525

This PR:
- refactors the redistribute implementation logic to make it more
sound, by figuring out the transform informations first and then apply
transformation step by step, we also cache the decisions so that it
could be reuse again
- for uneven sharding, refactor uneven sharding logic, and use a logical
  shape concept for each transform information to fix the uneven sharding
  multi-mesh redistribute bug

fixes https://github.com/pytorch/pytorch/issues/115310

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol committed:
[dtensor] refactor redistribute and fix uneven sharding redistribution

This PR:
- refactors the redistribute implementation logic to make it more
sound, instead of relying on things like decomp resharding logic, delete
that logic and replace that with inner sharding first redistribute.
- for uneven sharding, refactor uneven sharding logic, and use a logical
  dim size concept to fix the uneven sharding multi-mesh redistribute
bug

fixes https://github.com/pytorch/pytorch/issues/115310

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor redistribute and fix uneven sharding redistribution"


This PR:
- refactors the redistribute implementation logic to make it more
sound, instead of relying on things like decomp resharding logic, delete
that logic and replace that with inner sharding first redistribute.
- for uneven sharding, refactor uneven sharding logic, and use a logical
  dim size concept to fix the uneven sharding multi-mesh redistribute
bug

fixes https://github.com/pytorch/pytorch/issues/115310

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] refactor redistribute and fix uneven sharding redistribution"


This PR:
- refactors the redistribute implementation logic to make it more
sound, by figuring out the transform informations first and then apply
transformation step by step, we also cache the decisions so that it
could be reuse again
- for uneven sharding, refactor uneven sharding logic, and use a logical
  shape concept for each transform information to fix the uneven sharding
  multi-mesh redistribute bug

fixes https://github.com/pytorch/pytorch/issues/115310

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -121,25 +125,22 @@ def _local_shard_size_on_dim(
-            max(
-                min(size_on_dim, full_chunk_size * (idx + 1)) - full_chunk_size * idx,
-                0,
+        if size_on_dim < shard_starting_idx:
```
I see. So the last chunk will be empty in this case.

XilunWu reviewed (commented):
Reviewed the part except `redistribute.py` and it looks good! Will finish reviewing the rest later.

wz337 commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -1,75 +1,132 @@
-    Replicate from inner to outer dimension.
-    Shard from outer to inner dimension.
+    Replicate happens from inner to outer dimension.
+    Shard happens from outer to inner dimension.
```
Could we add a bit of comment regarding the expected behavior? 

Just want to confirm my understanding is right:
shard src -> partial or replicate => return negative mesh_dim
partial src and replicate src -> dst shard => return postive mesh_dim

And this helps because we need to do replication first, correct?

XilunWu replied:
agree. Since I have no background knowledge on `TransformInfo`, this function is not straightforward to people like me.

wanchaol replied:
> shard src -> partial or replicate => return negative mesh_dim
partial src and replicate src -> dst shard => return postive mesh_dim

> And this helps because we need to do replication first, correct?

Yes!

I did not change the original approach, but added some more comments for you to understand the details.

wz337 commented:
Seems the test failiure is relevant?

XilunWu commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -1,75 +1,132 @@
+    transform from source to target placement it might have multipl steps, i.e. it might
+    decompose Si -> Sj into Si -> R -> Sj.
+    This would detects if there're mis-aligned shardings between src/dst placements.
+    i.e. (Shard(0), Shard(0)) -> (Replicate(), Shard(0)), in this case Shard(0) -> Shard(0)
```
I think we can add an extra module in future to address this reshard algorithm if we haven't got one yet.

wanchaol replied:
No I don't think we should add an "extra" module. `redistribute.py` is already the module that handles redistribution, adding an extra module would make the system more complicated and it's not necessary. 

> It cannot generate the optimal resharding for case (Shard(0), Shard(0)) -> (Replicate(), Shard(0))

what do you mean by "allgather along the second mesh dim"? I think the current logic already gives the best and the valid resharding for this case.

XilunWu commented on torch/distributed/_tensor/redistribute.py:
```diff
@@ -146,18 +202,24 @@ def redistribute_local_tensor(
-                # decompose Shard(0) -> Shard(1) into Shard(0) -> Replicate -> Shard(1)
+                # NOTE: we don't support this case efficiently yet, the fallback path we are going here is
+                # to decompose Shard(0) -> Shard(1) into Shard(0) -> Replicate -> Shard(1)
+                # TODO: enable this with all_to_all
```
The 1-d mesh case seems straightforward and we had an implementation. The tricky case is 2-d and more. But I think we can start with the simple 1-d case first. WDYT?

wanchaol replied:
Hmmm I think we should not try to enable the easier case and then complicates too much on the redistribute logic, as I couldn't see a clean way to ONLY enable 1D mesh case and throw error for 2D or more. I want to avoid the trap that we enabled 1d but later we could not cleanly handle the multi mesh case. Let's try to enable the alltoall cleanly :) This refactor is trying to make redistribute simpler and could possibly enable for a multi mesh alltoall later.

In addition, the implementation we had probably aren't easily usable anymore as DTensor evolves to use functional collectives, we need to rewrote that with functional collective instead.

XilunWu reviewed (approved):
Reviewed `redistribute.py` part and I feel this PR looks good. Left some questions to help me better understand the context. Good job!

p.s. need to check the test failure as Iris said. The error msg is `ValueError: tensor_meta is not set`.

wanchaol committed:
Update on "[dtensor] refactor redistribute and fix uneven sharding redistribution"


This PR:
- refactors the redistribute implementation logic to make it more
sound, by figuring out the transform informations first and then apply
transformation step by step, we also cache the decisions so that it
could be reuse again
- for uneven sharding, refactor uneven sharding logic, and use a logical
  shape concept for each transform information to fix the uneven sharding
  multi-mesh redistribute bug

fixes https://github.com/pytorch/pytorch/issues/115310

[ghstack-poisoned]

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -121,25 +125,22 @@ def _local_shard_size_on_dim(
         # Compute the chunk size inline with ``torch.chunk``
+        if size_on_dim % num_chunks == 0:
+            full_chunk_size = size_on_dim // num_chunks
+            return full_chunk_size, full_chunk_size * rank if return_offset else -1
```
I tried to keep the original semantics, we can discuss later about whether we always want to return the offset, I think we probably should not always return offset as it requires some calculation on CPU and during runtime we want to save it

wanchaol commented:
> p.s. need to check the test failure as Iris said. The error msg is `ValueError: tensor_meta is not set`.

fixed the issue with test failures, it was due to the from_local backward redistribute missing the tensor meta.

awgu commented:
This is great! I think this will be helpful for some of my per-parameter FSDP unit testing. For 2D sharding and uneven case, I did not know of an easy way to test if the values are expected.

wanchaol commented:
@pytorchbot merge

XilunWu commented:
This PR helps fix issue #117728

wz337 commented:
> This is great! I think this will be helpful for some of my per-parameter FSDP unit testing. For 2D sharding and uneven case, I did not know of an easy way to test if the values are expected.

We have some existing tests for current flat param FSDP + TP. I have disabled the tests for uneven sharding case previously due to the issue. I can bring those tests back.
